homeassistant/components/sensor/repetier.py

```diff
@@ -0,0 +1,201 @@
+from homeassistant.util import Throttle
+
+REQUIREMENTS = ['pyrepetier==1.4.1']
+
+__version__ = '0.5.1'
+
+_LOGGER = logging.getLogger(__name__)
+
+SCAN_INTERVAL = timedelta(seconds=5)
+
+PLATFORM_SCHEMA = PLATFORM_SCHEMA.extend({
+    vol.Required(CONF_URL): cv.string,
+    vol.Required(CONF_API_KEY): cv.string,
+    vol.Optional(CONF_PORT, default='3344'): cv.string,
+    vol.Optional('decimals', default=1): int,
+    vol.Optional('state_percent', default=False): bool,
+})
+
+
+def parse_repetier_api_response(response):
+    """Parse the Repetier Server API json response."""
+    data_dict = {}
+
+    i = 0
+    for printer in response['data']:
+        _key = i
+        i += 1
+        if _key not in data_dict.keys():
+            data_dict[_key] = printer
+        else:
+            data_dict[_key].update(printer)
+    return data_dict
+
+
+def setup_platform(hass, config, add_devices, discovery_info=None):
+    """Set up the Repetier sensors."""
+    import pyrepetier
+    try:
+        data = pyrepetier.RepetierData(
+            parse_repetier_api_response,
+            url=config.get(CONF_URL),
+            port=config.get(CONF_PORT),
+            api_key=config.get(CONF_API_KEY))
+        data.update()
+        sensors = []
+        for key in data.data.keys():
+            sensors.append(RepetierSensor(config, key, data))
+        add_devices(sensors, True)
+    except (RuntimeError, TypeError, NameError):
+        _LOGGER.warning("Cannot setup Repetier sensors, check your config")
+
+
+class RepetierSensor(Entity):
+    """Class to hold Repetier Sensor basic info."""
+
+    def __init__(self, config, repetier_id, data):
+        """Initialize the sensor object."""
+        self._repetier_id = repetier_id
+        self._data = data
+        self._icon = 'mdi:printer-3d'
+        self._name = None
+        self._state = STATE_UNKNOWN
+        self._units = None
+        self._attributes = None
+        self._decimals = config.get('decimals')
+        self._show_pct = config.get('state_percent')
+        self.format_data()
+
+    @property
+    def name(self):
+        """Return the name of the sensor."""
+        return self._name
+
+    @property
+    def state(self):
+        """Return the state of the sensor."""
+        return self._state
+
+    @property
+    def icon(self):
+        """Icon to use in the frontend, if any."""
+        return self._icon
+
+    @property
+    def unit_of_measurement(self):
+        """Set units of measurement."""
+        return self._units
+
+    @property
+    def device_state_attributes(self):
+        """Attributes."""
+        return self._attributes
+
+    @Throttle(SCAN_INTERVAL)
+    def update(self):
+        """Update the sensor."""
+        try:
+            self._data.update()
+            self.format_data()
+        except (RuntimeError, TypeError, NameError):
+            _LOGGER.error("Error updating Repetier Server sensors")
+
+    def format_data(self):
```

Comment:
Instead of this we should implement sensor types, so that each data key is one sensor type. Use `CONF_MONITORED_CONDITIONS` to configure sensor types. See other sensor platforms that use this for examples.

Reply:
I'm not sure I follow you here - the module creates one sensor for each device, each sensor have the device status a attributes.
Why would you want every attribute moved into a seperate sensor?

Reply:
We prefer many individual sensors above sensors with loads of attributes.

Reply:
Where in the docs is this described?
I must say I strongly disagree on this matter.

Reply:
Compare with the octoprint component and platforms. Try to model if after that.
https://github.com/home-assistant/home-assistant/blob/dev/homeassistant/components/octoprint.py